src/peergos/shared/user/fs/TagIndexingService.java

```diff
@@ -0,0 +1,95 @@
+package peergos.shared.user.fs;
+
+import peergos.shared.Crypto;
+import peergos.shared.NetworkAccess;
+import peergos.shared.storage.auth.Bat;
+import peergos.shared.user.*;
+import peergos.shared.user.fs.transaction.Transaction;
+import peergos.shared.user.fs.transaction.TransactionService;
+import peergos.shared.util.ProgressConsumer;
+import peergos.shared.util.Serialize;
+
+import java.nio.file.Paths;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+public class TagIndexingService {
+    private static ProgressConsumer<Long> VOID_PROGRESS = l -> {
+    };
+    public static final String TAGS_INDEX_FILENAME = ".tagsindex";
+
+    private final FileWrapperUpdater tagsIndexDirUpdater;
+    private final FileWrapper tagsIndexHomeDir;
+    private final TransactionService transactions;
+    private final NetworkAccess networkAccess;
+    private final Crypto crypto;
+
+    public TagIndexingService(NetworkAccess networkAccess,
+                              TransactionService transactions,
+                              Crypto crypto,
+                              FileWrapper tagsIndexDir) {
+        this.tagsIndexDirUpdater = v -> tagsIndexDir.getUpdated(v, networkAccess);
+        this.tagsIndexHomeDir = tagsIndexDir;
+        this.transactions = transactions;
+        this.networkAccess = networkAccess;
+        this.crypto = crypto;
+    }
+
+    public CompletableFuture<TagIndex> addTagIndexEntries(TagIndex tagIndex, Map<String, Set<String>> entries) {
+        return CompletableFuture.completedFuture(tagIndex.addEntries(entries));
+    }
+
+    public CompletableFuture<TagIndex> removeTagIndexEntries(TagIndex tagIndex,  Map<String, Set<String>> entries) {
+        return CompletableFuture.completedFuture(tagIndex.removeEntries(entries));
+    }
+
+    public CompletableFuture<Snapshot> modifyTagIndexEntries(Snapshot version, Committer committer,
+                                                             Map<String, Set<String>> entriesToAdd,
+                                                             Map<String, Set<String>> entriesToRemove) {
+        return getTagIndexEntries(version).thenCompose(originalTagIndex -> removeTagIndexEntries(originalTagIndex, entriesToRemove))
+                .thenCompose(tagIndexAfterRemoval -> addTagIndexEntries(tagIndexAfterRemoval, entriesToAdd))
+                .thenCompose(tagIndexAfterAdding -> modifyTagIndexEntries(version, committer, tagIndexAfterAdding));
+    }
+
+    private CompletableFuture<Snapshot> modifyTagIndexEntries(Snapshot s, Committer committer, TagIndex newIndex) {
```

Comment:
🟢 
Can we change the name like `s` to something more meaningful (this comment applies to the whole file - there are variables like e.g. `v` or `x`)?

Reply:
Done